### ISSUE TITLE
Option "serialization" on p:output for atomic steps

### ISSUE THREAD

**Issue by xml-project** [xproc/3.0-specification#428]
Currently we have:
```
<p:output
  port? = NCName
  sequence? = boolean
  primary? = boolean
  content-types? = ContentTypes
  expand-text? = boolean
  serialization? = map(xs:QName,xs:anyAtomicValue) />
```
for atomic steps, but I wonder what "serialization" is for.

> If the processor is not serializing (if, for example, the pipeline has been called from another pipeline), then the serialization must be ignored.

And since an atomic step is always called from another step, the serialization options are always ignored.

Is there any (non-propertary) use case for this? I could think of calling an atomic step from within a p:library, but this is not a standard feature of a processor.

**Comment by gimsieke** [comment]
I don’t think that your presupposition that an atomic step is always called from another step is justified. At least for Calabash, there’s nothing wrong with calling, for example, `-s p:xslt` on the command line. Then the serialization option, at least on the `result` port, is meaningful.

**Comment by xml-project** [comment]
According to the specs, an atomic step is ALWAYS called from another step. 
What you mentioned is what I called a proprietary use case in my comment. In MorganaXProc you can call an atomic step from a library, but there is no place in the specs saying either use case is required.

The point is what to answer if someone asks what "serialization" is for and you can not find an answer in the specs. Possible solutions:
(1) Say that the processor may execute atomic steps by a processor specific mechanism.
(2) Remove "serialization". This does not prevent the implementer from doing any magic, but it leaves no un-explained feature in the specs. If you want to set serialization parameters for an output port: Write a pipeline!

**Comment by xml-project** [comment]
@gimsieke Do you prefer to keep it and add a note saying it is used, when processors are able to call atomic steps?

**Comment by xml-project** [comment]
@eriksiegel 

> ButI don't understand: an atomic step is ALWAYS called from another step? There must always be a first step that is "called", "processed" or whatever.

Yes, but this step is a compound step and for compound steps serialization makes sense. But we do not say in any place, that e.g. p:add-attribute may be called outside a compound step.

**Comment by ndw** [comment]
I think that’s a bug. We used to have a p:serialization element to define serialization parameters. When we made it a map, we must have just stuck it on p:output generally.

Atomic steps don’t serialize their output. Even in XML Calabash where you can run a step “directly”, I believe that’s described as syntactic sugar for a p:pipeline containing a single step.

We could leave it, and say it only applies on pipeline outputs, but that might be confusing. We could put p:serialization back in and hang the parameters off it. But it seems nicer not to have to have that level of indirection, I think.

We could make p:output on p:declare-step different than p:output elsewhere, but I wonder if it’s worth the effort?

**Comment by xml-project** [comment]
May be we should take away "@serialization" from p:output all together, because it is also strange to have it on p:for-each etc.

What about adding an option "serialization" to p:declare-step which is "map(xs:string, map(*))" with the port name as key?

**Comment by xml-project** [comment]
I am not opposed to p:serialization, but I thought the map would be easier to understand and save an additional XProc element.

**Comment by xml-project** [comment]
@gimsieke Sorry, I fail to see, what you are suggesting as a solution. Would you mind to elaborate on your point a bit more, so I can see, what syntax and implementation is suggested.

**Comment by gimsieke** [comment]
I first suggested that we allow a serialization map on `p:option` if it is a child of a `p:declare-step` that has a subpipeline, otherwise not. 

In addition, I suggested that we clarify the spec wrt the definition of atomic steps and direct invocation of atomic steps.

**Comment by gimsieke** [comment]
I’d prefer `p:output` with three slightly different models over re-introduction of `p:serialization`. I don’t think that there will be much confusion, in particular because these differences can nicely be captured in our Relax NG schema, so tooling (nXML, oXygen, …) will be informed to offer context-dependent completion only for the `p:output` attributes that are valid according to the schema.

**Comment by eriksiegel** [comment]
Agree with @gimsieke, forget my former suggestion about re-instating p:serialization.

**Comment by gimsieke** [comment]
Because @xml-project asked about RNG-based content completion, I created a small demonstration that oXygen will provide decent completion if configured only by our RNG (or RNC) schema. Achim was afraid that it wouldn’t even offer the serialization attribute unless and until there is a subpipline further down in `p:declare-step`. I confirmed that the serialization attribute will indeed be suggested in the absence of a subpipeline. A rather helpful error message will display when the `@serialization` attribute has been entered and the subpipeline is still missing. I think it’s quite usable and in no way confusing for the user to have 3 different, context-dependent models for `p:output` (2 of which I have sketched in attached [atomic.zip](https://github.com/xproc/3.0-specification/files/2133977/atomic.zip)). Likewise, nXML mode should be able to use the RNC variant for content completion.
